design-proposals/flavors-and-preferences.md

```diff
@@ -0,0 +1,482 @@
+
+## Repos
+
+* [kubevirt/kubevirt](https://github.com/kubevirt/kubevirt)
+
+# Design
+
+As stated above the ultimate goal of this work is to provide the end user with a simple set of choices when initially defining a VM within KubeVirt.
+
+To achieve this the existing [`VirtualMachineFlavor`](http://kubevirt.io/api-reference/main/definitions.html#_v1alpha1_virtualmachineflavor) CRDs will be heavily modified and extended to better encapsulate any resource, performance or schedulable attributes of a VM. A new set of `VirtualMachinePreference` CRDs will then be introduced to define any remaining attributes related to ensuring the selected guestOS can run.
+
+* Remove [`VirtualMachineFlavorProfile`](http://kubevirt.io/api-reference/main/definitions.html#_v1alpha1_virtualmachineflavorprofile) from [`VirtualMachineFlavor`](http://kubevirt.io/api-reference/main/definitions.html#_v1alpha1_virtualmachineflavor) and [`VirtualMachineClusterFlavor`](http://kubevirt.io/api-reference/main/definitions.html#_v1alpha1_virtualmachineclusterflavor).
+
+* Introduce a new `Spec` attribute and new `VirtualMachineFlavorSpec` type  itself with `CPU`, `Memory`, `GPUs`, `HostDevices` and `DeviceDefaults` attributes to [`VirtualMachineFlavor`](http://kubevirt.io/api-reference/main/definitions.html#_v1alpha1_virtualmachineflavor) and [`VirtualMachineClusterFlavor`](http://kubevirt.io/api-reference/main/definitions.html#_v1alpha1_virtualmachineclusterflavor).
+
+* Introduce new `VirtualMachinePreference` and `VirtualMachineClusterPreference` CRDs to the Flavors API and `VirtualMachinePreferenceSpec` type to define guestOS related preferences such as preferred device buses and models.
+
+* Allow for `CPU` preferred topology policy to be defined within `VirtualMachinePreferenceSpec` through a `PreferredTopology` attribute.
+
+* Extend [`VirtualMachineSpec`](http://kubevirt.io/api-reference/main/definitions.html#_v1_virtualmachinespec) to allow an optional reference to `VirtualMachinePreference` or `VirtualMachineClusterPreference`.
+
+## [`DomainSpec`](http://kubevirt.io/api-reference/main/definitions.html#_v1_domainspec) mappings to `VirtualMachinePreferenceSpec` and `VirtualMachineFlavorSpec`
+
+The following basic rules are applied when deciding which attributes of the [`DomainSpec`](http://kubevirt.io/api-reference/main/definitions.html#_v1_domainspec) will be mapped to the new `VirtualMachinePreferenceSpec` and `VirtualMachineFlavorSpec` types:
+
+* Resource, schedulable and performance related attributes of the DomainSpec reside within `VirtualMachineFlavorSpec`
+
+* Guest OS related Preference reside within `VirtualMachinePreferenceSpec`
```

Comment:
Not sure where the comment belongs, but: Because it's preferences, I assume that a user is able to define some domain value in VMI.spec.domain or rather VM.spec.template.spec.domain, like autoAttachGraphicalDisplay to set his custom specified value.
If this is not set, then the value from the preference would be taken.

Is this assumption correct?

Reply:
ACK yeah I'll expand the initial design section to contain this and some related points, thanks!